File: app/src/main/java/com/hulkdx/moneymanager/injection/module/ApplicationModule.java
/**
 * Created by Mohammad Jafarzadeh Rezvan on 7/6/2017.
 */

package com.hulkdx.moneymanager.injection.module;

import android.app.Application;
import android.content.Context;
import dagger.Module;
import dagger.Provides;
import com.hulkdx.moneymanager.injection.ApplicationContext;

/**
 * Provide application-level dependencies.
 */
@Module
public class ApplicationModule {
    protected final Application mApplication;

    public ApplicationModule(Application application) {
        mApplication = application;
    }

    @Provides
    Application provideApplication() {
        return mApplication;
    }

    @Provides
    @ApplicationContext
    Context provideContext() {
        return mApplication;
    }

}

File: presentation/src/main/java/com/hulkdx/moneymanagerv2/di/modules/ApplicationModule.kt
package com.hulkdx.moneymanagerv2.di.modules

import dagger.Module
import dagger.Provides
import hulkdx.com.domain.di.BackgroundScheduler
import hulkdx.com.domain.di.UiScheduler
import hulkdx.com.domain.util.CustomThreadExecutor
import io.reactivex.Scheduler
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers

/**
 * Created by Mohammad Jafarzadeh Rezvan on 09/11/2018.
 */
@Module
abstract class ApplicationModule {
    @Module
    companion object {
        @JvmStatic
        @UiScheduler
        @Provides
        fun provideUiScheduler(): Scheduler = AndroidSchedulers.mainThread()

        @JvmStatic
        @BackgroundScheduler
        @Provides
        fun provideBackgroundScheduler(customThreadExecutor: CustomThreadExecutor): Scheduler =
                Schedulers.from(customThreadExecutor)
    }
}
